storage/block/wired_list.go

```diff
@@ -0,0 +1,269 @@
+//
+// The WiredList itself does not allocate a per-entry datastructure to keep track
+// of what is active and what is not. Instead, it creates a "virtual list" ontop
+// of the existing blocks that are in memory by manipulating struct-level pointers
+// on the DatabaseBlocks which are "owned" by the list. In other words, the
+// DatabaseBlocks are scattered among numerous namespaces/shards/series, but they
+// existed in virtual sorted order via the prev/next pointers they contain, but
+// which are only manipulated by the WiredList.
+//
+// The WiredList ONLY keeps track of blocks that are read from disk. Blocks that
+// are created by rotating recently-written data out of buffers and into new
+// DatabaseBlocks are managed by the background ticks of the series. The background
+// tick will avoid closing blocks that were read from disk, and a block will never
+// be provided to the WiredList if it wasn't read from disk. This prevents tricky
+// ownership semantics where both the background tick and and the WiredList are
+// competing for ownership / trying to close the same blocks.
+
+package block
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/m3db/m3db/clock"
+	"github.com/m3db/m3db/runtime"
+	"github.com/m3db/m3x/instrument"
+
+	"github.com/uber-go/tally"
+)
+
+const (
+	wiredListEventsChannelLength = 65536
+	wiredListSampleGaugesEvery   = 100
+)
+
+// WiredList is a database block wired list.
+type WiredList struct {
+	sync.Mutex
+
+	nowFn clock.NowFn
+
+	// Max wired blocks, must use atomic store and load to access.
+	maxWired int64
+
+	root      dbBlock
+	length    int
+	updatesCh chan DatabaseBlock
+	doneCh    chan struct{}
+
+	metrics wiredListMetrics
+}
+
+type wiredListMetrics struct {
+	unwireable           tally.Gauge
+	limit                tally.Gauge
+	evicted              tally.Counter
+	pushedBack           tally.Counter
+	inserted             tally.Counter
+	evictedAfterDuration tally.Timer
+}
+
+// NewWiredList returns a new database block wired list.
+func NewWiredList(
+	runtimeOptsMgr runtime.OptionsManager,
+	iopts instrument.Options,
+	copts clock.Options,
+) *WiredList {
+	scope := iopts.MetricsScope().
+		SubScope("wired-list")
+	l := &WiredList{
+		nowFn: copts.NowFn(),
+		metrics: wiredListMetrics{
+			// Keeps track of how many blocks are in the list
+			unwireable: scope.Gauge("unwireable"),
+			limit:      scope.Gauge("limit"),
+			// Incremented when a block is evicted
+			evicted: scope.Counter("evicted"),
+			// Incremented when a block is "pushed back" in the list, I.E
+			// it was already in the list
+			pushedBack: scope.Counter("pushed-back"),
+			// Incremented when a block is inserted into the list, I.E
+			// it wasn't already present
+			inserted: scope.Counter("inserted"),
+			// Measure how much time blocks spend in the list before being evicted
+			evictedAfterDuration: scope.Timer("evicted-after-duration"),
+		},
```

Comment:
Just for consistency, might be worth creating a `newWiredListMetrics(...) wiredListMetrics { return /* ... */ }` just to follow the pattern of the other components.

Reply:
done